### PR TITLE
Single YAML-based Azure DevOps build

### PR DESCRIPTION
**Pull request by jnm2**
I had some experience with Azure DevOps recently and thought I'd put it to the test and fix some problems.

➡ Fixes the issue that our current three builds don't report failures to GitHub checks.

➡ Having three separate builds means we have to change settings in three places.

➡ The tests across the OSs can be viewed together.

➡ Source control gives us history and allows us to evolve the build configuration along with other files in the repository. See the build.cake change in the same commit as azure-piplines.yml, for example.

This is what this build definition looks like, live:
https://nunit.visualstudio.com/NUnit/_build?definitionId=11

Here's the last build I did, which succeeded (validating this PR):
https://nunit.visualstudio.com/NUnit/_build/results?buildId=1152

Here's the one before that, which failed (see the Tests tab):
https://nunit.visualstudio.com/NUnit/_build/results?buildId=1151

The Windows build runs a little longer and saves the packages as a downloadable artifact. (Down the road, we might want to do a release pipeline. I figured out how to publish my VS extension to GitHub and the VS marketplace, tag and bump versions with a single click. It was pretty exciting once all the parts were in place.)

### Plan

Right now I have the build I set up in Azure DevOps set to override the trigger in the YAML so that this PR branch doesn't cause the new build to trigger for *everyone's* branches, which is what the YAML is intentionally specifying.
As soon as this PR is merged, I'll undo that and then **disable the three existing builds**. This way, all new branches and PRs will only trigger the new single build.

Does this sound good to everyone? /cc @nunit/framework-team

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
+trigger:
+  branches:
+    include: [ '*' ]
+    exclude: [ 'refs/tags/*' ]
```
Comment (jnm2) [inline azure-pipelines.yml]: This keeps it from triggering on tags, which I found surprising in another project where a release pipeline tagged the sources as the last stop of doing a release.

File: build.cake [hunk 2]
```diff
@@ -49,7 +49,7 @@ var NetCoreTests = new String[]
 //////////////////////////////////////////////////////////////////////
 
 var PROJECT_DIR = Context.Environment.WorkingDirectory.FullPath + "/";
-var PACKAGE_DIR = PROJECT_DIR + "package/";
+var PACKAGE_DIR = Argument("artifact-dir", PROJECT_DIR + "package") + "/";
```
Comment (jnm2) [inline build.cake]: Allows overriding the `package` path from the command line, which you can see here:
https://github.com/nunit/nunit/pull/3096/files#diff-fec826feae04e51c0d94076385408bdcR24

File: src/NUnitFramework/tests/Constraints/DelayedConstraintTests.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ public class DelayedConstraintTests : ConstraintTestBase
         // constructing the constraint class, and verify that they work.
 
         private const int DELAY = 100;
-        private const int AFTER = 300;
+        private const int AFTER = 500;
```
Comment (jnm2) [inline src/NUnitFramework/tests/Constraints/DelayedConstraintTests.cs]: Azure DevOps' macOS has been failing `ThatBlockingDelegateWhichSucceedsWithPolling_ReturnsEarly` on almost every build for a while. 300 ms is just too close to 100 ms for extremely shared servers.

**Comment by jnm2** [comment]
I think some of our test runs are overwriting each other. I copied the exact test result publishing that our current three ADO builds are doing, but I want to play with this. In theory we should have 18 runs uploaded in total, or actually 36 if none are merged but I want to merge them:

1. Test45 on Windows (merge framework and nunitlite result files since the test names will be unique between them)
2. Test40 on Windows (merge framework and nunitlite)
3. Test30 on Windows (merge framework and nunitlite)
4. Test20 on Windows (merge framework and nunitlite)
5. TestNetStandard14 on Windows (merge framework and nunitlite)
6. TestNetStandard20 on Windows (merge framework and nunitlite)
7. Test45 on Linux (merge framework and nunitlite)
8. Test40 on Linux (merge framework and nunitlite)
9. Test30 on Linux (merge framework and nunitlite)
10. Test20 on Linux (merge framework and nunitlite)
11. TestNetStandard14 on Linux (merge framework and nunitlite)
12. TestNetStandard20 on Linux (merge framework and nunitlite)
13. Test45 on macOS (merge framework and nunitlite)
14. Test40 on macOS (merge framework and nunitlite)
15. Test30 on macOS (merge framework and nunitlite)
16. Test20 on macOS (merge framework and nunitlite)
17. TestNetStandard14 on macOS (merge framework and nunitlite)
18. TestNetStandard20 on macOS (merge framework and nunitlite)

**Comment by mikkelbu** [comment]
I like this, especially the part about being able to see the TestResults from the different test runs (something I've lacked several times)

**Comment by jnm2** [comment]
Because I didn't feel like repeating all seven lines of the 'Publish test results' task six times and having to update the YAML if we start doing different framework targets, I put everything under `\test-results\{framework}\{assembly name}.xml` and wrote this instead:

```yml
  - powershell: |
      if (Test-Path 'test-results') {
          foreach ($frameworkDir in Get-ChildItem -dir 'test-results') {
              $resultFiles = Get-ChildItem -file "test-results\$frameworkDir" -Filter *.xml
              $quotedAndJoinedFilePaths = ($resultFiles | ForEach-Object { "'" + $_.FullName + "'" }) -join ';'
              Write-Host "##vso[results.publish type=NUnit;mergeResults=true;runTitle='$frameworkDir/$env:AGENT_JOBNAME';publishRunAttachments=true;resultFiles=$quotedAndJoinedFilePaths]"
          }
      }
    displayName: Publish test results
    condition: succeededOrFailed()
```

Now I'll have to figure out the Bash equivalent. I wonder if I should just put this in the Cake script. I like the idea of making build.cake agnostic of a particular CI vendor.

**Comment by jnm2** [comment]
I'm going to do it with Cake. The upside is that tests become available as each of the 18 runs finishes.

**Comment by jnm2** [comment]
Last thing before we can merge: I'm hitting https://github.com/cake-build/cake/issues/2337.

**Comment by jnm2** [comment]
So this is cool: you can watch each of the 18 test results rolling in one at a time for the various frameworks and OSs. Not sure why it seemed to wait till the end before; maybe a recent feature and I'm hitting updated VMs now?

**Comment by jnm2** [comment]
Can't figure this out. If I run a build with diagnostic logs, the test results all upload fine. If I don't, the Windows ones never get uploaded.

**Comment by jnm2** [comment]
Uploading test results only happens if the jobs happen upon the correct agents. When it works, it says "Agent: Hosted VS2017 2." When it fails, it says "Agent: Hosted Agent." Similar for Linux and macOS.

I wonder if this is what's going on:

https://github.com/Microsoft/azure-pipelines-tasks/blob/f8526d84cde186125cc589433fb950d437d1ef77/Tasks/PublishTestResultsV2/task.json#L20

> ```json
> "minimumAgentVersion": "2.0.0"
> ```

Now that I'm not using `task: PublishTestResults@2` in the YAML, it's not requiring the capable agents.

**Comment by jnm2** [comment]
Problems with the automated approach summarized at https://github.com/nunit/nunit/issues/3012#issuecomment-441517922. (My wife says this makes her happy she is not a software developer. 😆)

For now, I'll put the 18 copies of the 'Publish test results' YAML into `azure-pipelines.yml` just so we can move forward.

**Comment by jnm2** [comment]
Whew, what a good thing to see! I am not about to admit how many of the previous 24 hours I've spent on this. 😅

image


Review and merge at will!

(To reiterate from the top: As soon as this is merged, unless someone has a concern, I'll stop https://nunit.visualstudio.com/NUnit/_build?definitionId=11 from overriding the YAML by doing manual builds only, and then disable the three existing builds.)

**Comment by jnm2** [comment]
Resolved merge conflicts and copied a commit from https://github.com/nunit/nunit/pull/3095 to fix a problem with `master`.

**Comment by jnm2** [comment]
Resolved merge conflicts.

**Comment by jnm2** [comment]
> You'll update the build definition once this is merged.

Done, and I see you removed the previous builds. 👍

**Comment by rprouse** [comment]
PR merged. I've deleted the three platform specific builds but I don't see the YAML override in the trigger. Just came in, you did it 😄 

We should make this a Required build. Do you know if the build look at the version of the YML file in master, or we need to wait a bit so that existing PRs are merged. I'm worried that existing PRs won't have the YML file so won't get built if we set it as required.

**Comment by jnm2** [comment]
My push to https://github.com/nunit/nunit/pull/3095 didn't trigger it, so I'm trying to figure out why not. I'm also going cold turkey with COLEMAK right now, so I'm a little slow. Lol.

**Comment by jnm2** [comment]
> Do you know if the build look at the version of the YML file in master, or we need to wait a bit so that existing PRs are merged.

I'm actually not sure.

**Comment by rprouse** [comment]
Let's hold off on making it required until we know then. Also ignore my comment on your fork pr, I saw it before your comments here.

**Pull request merged**